### ISSUE TITLE
Reference new project in README.md

### ISSUE THREAD

**Issue by asieira** [afilipovich/gglsbl#19]
I have written a new project that basically implements a dockerized REST wrapper around gglsbl and published it at https://github.com/mlsecproject/gglsbl-rest if you are interested.

Any suggestions and comments on the implementation will be most welcome. And a link to this repository from gglsbl's README.md would be much appreciated. :)

**Comment by asieira** [comment]
https://twitter.com/AlexandreSieira/status/871844131596754944

**Comment by afilipovich** [comment]
Thanks for sharing that useful addition. Added a mention to README.
Wonder what was rationale for having two DBs for caching? Did you experience any issues with running GSB queries while DB is being updated? I presumed that SQLite transaction isolation permits running concurrent reads and updates.

**Comment by asieira** [comment]
I was in a hurry, read https://sqlite.org/faq.html#q5 and was concerned that the writes might cause downtime to the service. Plus, I imagined there could be issues for any requests being handled mid-update due to the database being in an inconsistent state.

If you assure me any or all of this is wrong, I could very well test using a single database file and see what happens. Did you ever test having multiple lookups running at the same time as an update? Is this something you would expect to work normally?

**Comment by afilipovich** [comment]
I see, thanks. Based on what you say, service downtime may be possible. However DB state should be always consistent as update runs in a transaction. I would not bet my life on that but would not mind relying on it in production.

Lookups may perform writes as well (to cache full hash values) but that happens in less than 1% of queries. So far I have not observed any issues with concurrent lookup.

**Comment by asieira** [comment]
Quick question before I close this issue... do you think it would be feasible to have gglsbl use a key-value store (such as DynamoDB or redis) as a naturally shared storage for multiple containers, instead of local sqlite databases?

If that was the case and the update processes could be made in a way to keep the data consistent at all times, I believe this would be the ideal approach to something like gglsbl-rest.

**Comment by afilipovich** [comment]
I have considered using Redis but did not find how to address the same data entry using multiple keys efficiently. E.g. be able to get all entries with hash prefix P while being able to delete all the records within hash prefix list L.

In theory this can be done by storing each hash prefix entry with unique (primary) key and maintaining index hashes, e.g. {hash_prefix_value -> hash_prefix_entry_primary_key}, but in practice it would be too slow when there are 100MM entries to insert (it would require several Redis operations per entry).

I think Redis can be used in combination with SQLite. The workflow would be:
- run partial hash prefix cache update as it is now using SQLite
- after each update upload SQLite state to Redis
- use Redis for lookup queries

But this is probably more complicated than your approach with lightweight REST service.

**Comment by Stefan-Code** [comment]
@afilipovich If you need more flexibilty regarding query operations, you might want to have a look at MongoDB instead of Redis (although I'm not implying that's a good idea here, I don't know how performance will compare to SQLite).

In SQL right now you are primary keying `value, threat_type, platform_type, threat_entry_type`.
AFAIK with Redis, you would need a seperate DB per threat list, and store the hashes for each independently. It should work without index hashes.

**Comment by afilipovich** [comment]
@Stefan-Code , I emailed you several weeks ago regarding CI, perhaps I sent it to wrong email or the message ended up in spam.

I don't see why MongoDB would be much faster than, say, Postgres which I tried.
Storing each threat_type/platform_type/threat_entry_type permutation in a separate Redis hash map would require checking each of these maps when performing lookup. Perhaps it still would be fast comparing to SQLite, but it looks like a long shot to me. 

IMHO running gglsbl as a micro-service, as @asiera did, is a good trade off between simplicity and scalability.

**Comment by Stefan-Code** [comment]
@afilipovich I just went through my spam folder, nothing there. Did you try: stefan[dot]github[at]gmail.com

**Comment by afilipovich** [comment]
Yep that email address, re-sent it once again from aleh@vaolix.com.

**Comment by asieira** [comment]
In DynamoDB you have the concept of a partition key and a sort key (http://docs.aws.amazon.com/amazondynamodb/latest/developerguide/HowItWorks.CoreComponents.html#HowItWorks.CoreComponents.PrimaryKey) that you can use to store the partial and full hashes of an entry. Using that schema, you could very easily list all entries in the table that match a specific partial hash. You can also have secondary indexes (http://docs.aws.amazon.com/amazondynamodb/latest/developerguide/LSI.html) on other fields.

The reason why I'm bringing this up is that I was hoping to avoid each container independently downloading and locally storing the database in its filesystem, which is what gglsbl-rest currently does. I assume a key/value store would outperform a relational database such as PostgreSQL for the sort of straight key / value lookup that's required for a gglsbl lookup operation, though of course I might be wrong. :)

**Comment by afilipovich** [comment]
From what I understand DynamoDB is specific to Amazon services. I would not like to tie the generic library code to specific service provider.

Perhaps I do not fully understand your use case... why would you prefer running external database server providing data backend to gglsbl instead of running gglsbl-rest micro-service?

**Comment by asieira** [comment]
That is true, DynamoDB is specific to AWS. So maybe the solution would be to allow for different persistence implementations on the client, and provide alternatives such as relational database and DynamoDB in addition to sqlite.

Basically I was trying to go for a model in which there would be a shared persistent storage of the database across gglsbl-rest containers, instead of each container independently downloading and storing a local copy.

The main reason for that is autoscaling. See, a new gglsbl-rest container needs to perform an initial full database download when it starts, and this can take 5-10 minutes. The typical autoscaling scenario goes "if there are too many requests for the existing container pool, start more containers to share the load as quickly as possible". Only that typically won't work because by the time the new containers actually start serving requests the damage is already done and the clients will have errored out.

I remember from another ticket that you tried using a relational database for storage and that there was a significant performance hit. Which is why I brought up using a K/V store, which is typically much faster for the kind of operation gglsbl does. Maybe using that would be an acceptable middle ground between sqlite and a relational database.

**Comment by asieira** [comment]
Speaking of Redis, might the Composite Indexes idea described here be a solution for searching for all entries that have the same hash prefix? https://redis.io/topics/indexes

**Comment by afilipovich** [comment]
@asieira, thanks for the ideas!
My concern is that full-blown database would be the bottleneck anyway and would be difficult to scale, while performance hit would be significant.

With SQLite file stored on RAMdisk (/dev/shm or /tmp mounted on tmpfs) I got 5-20 times higher performance comparing to Postgres running on the same machine.

Probably hybrid solution of SQLite + Redis would be better suited for distributed applications requiring high throughput. The workflow would be like:

1. Use SQLite to resolve what hash prefixes need to be added/deleted in the local cache.
2. After sync finishes copy all hash prefixes to Redis.
3. When performing lookup, check Redis for hash prefix.
4. If hash prefix was found, perform full hash lookup and cache its result in Redis as well as SQLite.

In that scenario only one server will be required to facilitate local cache update and main brunt of the lookup will be on Redis which is easier to scale.


The disadvantage of this approach is complexity of the deployment as `gglsbl` will be split into "lookup client" and "sync client" parts potentially running on different servers.

In the end, simply running a few redundant `gglsbl-rest` instances to handle load spikes might be a cheaper option.

What do you think?

**Comment by asieira** [comment]
I'm sorry I'm not as familiar as you are with how gglsbl works, so I'm still not exactly clear how the scenario you are describing would work.

Let's assume as you did that there are two types of containers. Multiple instance of "lookup" containers that are serving requests and performing URL lookups. And a single "update" container that is started every 30 minutes to update the shared database.

Questions:

* Would the "lookup" containers also need to use sqlite, or just the "update" container? 

* An initial reading of your scenario indicates to me that steps 1-2 happen on the "update" container and steps 3-4 on the "lookup" containers. But if that is the case, why would step 4 mention update the sqlite database? Wouldn't updating an entry with the URL as the key on Redis be enough?

**Comment by afilipovich** [comment]
Sorry for not being specific enough.
In that scenario:
- there will be single container with SQLite DB which syncs with Google API
- Redis running either in container or on AWS
- lookup client library which can be imported directly from application. It will only query Redis.
- I think you are right and there is no need to store full hash query result in SQLite, only Redis should be sufficient.

**Comment by asieira** [comment]
Perfect, that is a very interesting architecture exercise we just did. :) 

I believe this approach would be more amenable to autoscaling since when a lookup instance starts it can immediately start honoring requests using the Redis data.

How difficult do you think this would be to implement? I am more than willing to help out with the coding if you are up to it. :)

**Comment by afilipovich** [comment]
It should not be a lot of coding and I think it can be done in 4 to 8 hours. I will give it a try.

The biggest help would be verifying that it works correctly, measuring performance and writing a guide on how to deploy it in distributed environment.

**Comment by afilipovich** [comment]
@asieira, the code in "common_cache" branch uses both Sqlite to facilitate cache sync and Redis to share the cache between multiple servers.

The setup requires that worker that runs syncs is configured to use both Sqlite and Redis; servers which perform lookups should be configured to use only Redis (set db_path=None).

it took more work than I expected as there are multiple failure modes to consider.

Looking forward to know how/if it works for you.

**Comment by asieira** [comment]
Sorry I haven't given you any feedback on that yet, I have been tied up on business travel for quite some time recently. I'll try to test this out in the following weeks and let you know how it goes.

**Comment by afilipovich** [comment]
No problems, thanks!
